### PR TITLE
Initial plaso pubsub client code.

### PR DESCRIPTION
**Pull request by aarontp**
This is adapted from jberggren's initial code.  Will stick it in this tools/ dir for now and will merge functionality into the turbinia/ dir as I go.

### CODE REVIEW

File: tools/plaso_pubsub.py [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+# Turbinia
+from turbinia import config
+
+
+def GetCurrentTimeUTC():
+  return datetime.datetime.utcnow().strftime(u'%Y-%m-%dT%H:%M:%S+0000')
+
+
+def IsBlockDevice(path):
+  if not os.path.exists(path):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add docstring

File: tools/plaso_pubsub.py [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+    return
+  mode = os.stat(path).st_mode
+  return stat.S_ISBLK(mode)
+
+
+def WriteToStdOut(message):
+  # TODO(aarontp): Convert to standard logging mechanism
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add docstring

+1 on changing to python logging instead

File: tools/plaso_pubsub.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+  msg = u'{0:s} {1:s}\n'.format(GetCurrentTimeUTC(), message)
+  sys.stdout.write(msg)
+  sys.stdout.flush()
+
+
+def WriteToStdErr(message):
+  msg = u'{0:s} {1:s}\n'.format(GetCurrentTimeUTC(), message)
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add docstring

File: tools/plaso_pubsub.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+  sys.stderr.write(msg)
+  sys.stderr.flush()
+
+
+def CreateServiceClient(service):
+  """Creates an API client for talking to Google Cloud."""
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args to docstring

File: tools/plaso_pubsub.py [hunk 5]
```diff
@@ -0,0 +1,325 @@
...
+  WriteToStdOut(u'Create API client for service: {0:s}'.format(service))
+  credentials = oauth2client.GoogleCredentials.get_application_default()
+  http = httplib2.Http()
+  credentials.authorize(http)
+  return discovery.build(service, 'v1', http=http)
+
+
+class TimesketchApiClient(object):
+
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add docstring

File: tools/plaso_pubsub.py [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+                                                          sketch_id)
+    form_data = {u'timelines': [index_id]}
+    response = self.session.post(resource_url, json=form_data)
+    response_dict = response.json()
+    timeline_id = response_dict[u'objects'][0]['id']
+    return timeline_id
+
+
+class PlasoProcessor(object):
+
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add docstring

File: tools/plaso_pubsub.py [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+  def __init__(self):
+    config.LoadConfig()
+
+  def _WaitForOperation(self, client, operation):
+    """Wait for an operation to complete."""
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args

File: tools/plaso_pubsub.py [hunk 8]
```diff
@@ -0,0 +1,325 @@
...
+      WriteToStdOut(u'Status: {0:s}'.format(result[u'status']))
+      if result[u'status'] == u'DONE':
+        if u'error' in result:
+          WriteToStdErr(result)
+          raise RuntimeError(result[u'error'][u'errors'][0][u'message'])
+        return result
+      time.sleep(1)
+
+  def _AttachDisk(self, client, disk):
+    """Attaches a persistent disk to the machine."""
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args

File: tools/plaso_pubsub.py [hunk 9]
```diff
@@ -0,0 +1,325 @@
...
+    operation = client.instances().attachDisk(
+        instance=config.INSTANCE,
+        project=config.PROJECT,
+        zone=config.ZONE,
+        body={u'deviceName': config.DEVICE_NAME,
+              u'source': disk}).execute()
+    self._WaitForOperation(client, operation[u'name'])
+
+  def _DetachDisk(self, client):
+    """Detaches the disk from the machine."""
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args

File: tools/plaso_pubsub.py [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+    WriteToStdOut(u'Detaching disk')
+    operation = client.instances().detachDisk(
+        instance=config.INSTANCE,
+        project=config.PROJECT,
+        zone=config.ZONE,
+        deviceName=config.DEVICE_NAME).execute()
+    self._WaitForOperation(client, operation[u'name'])
+
+  def _CopyFileToBucket(self, client, filename, object_name):
+    WriteToStdOut(u'Uploading file: {0:s}'.format(filename))
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args and returns

File: tools/plaso_pubsub.py [hunk 11]
```diff
@@ -0,0 +1,325 @@
...
+        elif e.resp.status in [400]:
+          error_count += 1
+          if error_count > 5:
+            raise
+          time.sleep(5)
+        else:
+          raise
+    return True
+
+  def Process(self, persistent_disk_name, project_name, disk_name):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args

File: tools/plaso_pubsub.py [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    _RETRY_COUNT = 0
+    while _RETRY_COUNT < _RETRY_MAX:
+      if IsBlockDevice(path):
+        WriteToStdOut(u'Block device: OK')
+        break
+      if os.path.exists(path):
+        WriteToStdOut(u'Block device: Current mode is {0}'.format(
+            os.stat(path).st_mode))
+      _RETRY_COUNT += 1
+      time.sleep(1)
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Maybe use a constant here?
e.g. _SLEEP_TIME_SEC = 1

File: tools/plaso_pubsub.py [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+
+class TimesketchApiClient(object):
+
+  def __init__(self, server_ip, username, password):
+    self.server_ip = server_ip
+    self.host_url = u'https://{0:s}'.format(self.server_ip)
+    self.session = self._CreateSession(username, password)
+
+  def _CreateSession(self, username, password):
+    session = requests.Session()
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Add args

File: tools/plaso_pubsub.py [hunk 14]
```diff
@@ -0,0 +1,325 @@
...
+        'x-csrftoken': csrf_token,
+        'referer': self.server_ip
+    })
+
+    # Do a POST to the login handler to set up the session cookies
+    form_data = {u'username': username, u'password': password}
+    session.post(u'{0:s}/login/'.format(self.host_url), data=form_data)
+    return session
+
+  def CreateSketch(self, name, description):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Docstring and add args and returns

File: tools/plaso_pubsub.py [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+    resource_url = u'{0:s}/api/v1/sketches/'.format(self.host_url)
+    form_data = {u'name': name, u'description': description}
+    response = self.session.post(resource_url, json=form_data)
+    response_dict = response.json()
+    sketch_id = response_dict[u'objects'][0]['id']
+    return sketch_id
+
+  def UploadTimeline(self, timeline_name, plaso_storage_path):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Docstring and add args and returns

File: tools/plaso_pubsub.py [hunk 16]
```diff
@@ -0,0 +1,325 @@
...
+    resource_url = u'{0:s}/api/v1/upload/'.format(self.host_url)
+    files = {'file': open(plaso_storage_path, 'rb')}
+    form_data = {u'name': timeline_name}
+    response = self.session.post(resource_url, files=files, data=form_data)
+    response_dict = response.json()
+    index_id = response_dict[u'objects'][0]['id']
+    return index_id
+
+  def AddTimelineToSketch(self, sketch_id, index_id):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: Docstring and add args and returns

File: tools/plaso_pubsub.py [hunk 17]
```diff
@@ -0,0 +1,325 @@
...
+    """Detaches the disk from the machine."""
+    WriteToStdOut(u'Detaching disk')
+    operation = client.instances().detachDisk(
+        instance=config.INSTANCE,
+        project=config.PROJECT,
+        zone=config.ZONE,
+        deviceName=config.DEVICE_NAME).execute()
+    self._WaitForOperation(client, operation[u'name'])
+
+  def _CopyFileToBucket(self, client, filename, object_name):
```
Comment (berggren) [inline tools/plaso_pubsub.py]: I think we should move the GCE stuff out to its own class. But we can do that in a later CL. What do you think?

File: tools/plaso_pubsub.py (hunk 1 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 2 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 3 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 4 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 5 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 13 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 14 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 15 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 16 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 6 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 7 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 8 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 9 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 17 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Yeah, I'll be completely refactoring all of this in future CLs.

File: tools/plaso_pubsub.py (hunk 10 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 11 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: Done.

File: tools/plaso_pubsub.py (hunk 12 above)
Comment (aarontp) [inline tools/plaso_pubsub.py]: I added one since it's easy, but this will probably be refactored out later.

**Comment by aarontp** [comment]
Done.

**Pull request merged**